include/VaultClient.h

```diff
@@ -8,6 +8,7 @@
 
 #include <unordered_map>
 #include <curl/curl.h>
+#include <json.hpp>
```

Comment:
Since this header is exported as part of the library, including `json.hpp` here forces a dependency that isn't packaged, and ultimately goes against the grain of the goals of this library to be externally agnostic with regard to json parsing. This will need to be removed before we can merge this PR.

Reply:
If you take a look at the other places where the library wants parameters, a container is passed and iterated through to produce the JSON output.

Reply:
Yes, I have seen the Parameters structure used. I have used it where parameters follow strictly the string key:string value format. 

However in some calls I feel it is a bit too restrictive, for example if you want to pass a user name when creating a token. But in other cases where you need to use list parameters ok std::pairs as values too.

I understand that you want to stay agnostic of json parsing. I am not sure however what other options is. 

I have seen the usage of JsonProducer in post request for example in KeyValue store setting, but I found it very confusing, because the interface is hiding the fact that actually the parameters are going to the "data" field. When writing a call one has to check the http documentation of vault to find the right fields.

Sidenote : If it is about compilation there is a forward declare header in nlohmann too. include/nlohmann/json_fwd.hpp

Reply:
Hello, I've been watching this thread since I am interested in this functionality too. 

> However in some calls I feel it is a bit too restrictive, for example if you want to pass a user name when creating a token. But in other cases where you need to use list parameters ok std::pairs as values too.

I think if the map<string, string> has hit some hard limits as you mentioned, wanting to pass std::pair and other unstringy values through, maybe it's type to define some strongly typed c++ structs as parameters to these functions that need it (instead of JsonParameters)? That way we can satisfy @abedra's requirement to not link any other dependencies onto clients for the public interface.


For example

```c++

class CreateTokenParameters {
  std::vector<std::string> policies;
  std::map<std::string, std::string> meta;
  std::string ttl;
  bool renewable;
  ...
}

std::optional<std::string> createToken(const CreateTokenParameters &parameters);
    
```
This should also help with needing to constantly reference the documentation, as developers can have the benefit of autocomplete etc. for all the complex parameters. That's my two cents!

Reply:
Seems lik a good idea. To be honest I am not too opposed to Vault::Params. I just wanted to start a discussion.

If you want @abedra I can just change to param and be over with it.

Reply:
I tend to share @sundbry's thoughts here, and believe I've leaned that way in previous comments. I still think there's good reason to draw the line at exporting only the library interface and not forcing a JSON library on consumers. That really leaves us at making pods that can be serialized as the next logical choice. When I first started this library I realized that standard containers would eventually be outgrown and would have to shift to something better, but I wanted breadth and that would have caused a longer initial cycle. This case definitely seems like one of those where it's stretching the boundary both in terms of correctness as well as flexibility.

I would find it perfectly acceptable to deviate here and create pods to represent the different arguments to a call. The internal JSON library has some mechanisms built in to define serialization and deserialization.

Reply:
okay, if we define the parameters as a struct (let's say CreateTokenParameters):
 where should we put the definition?
for json serialization there are some nice macros, such as NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE() in the newer version. To use this we would need to upgrade nlhomann version, and also still would need. the include json.hpp in VaultClient.h if we keep the struct definitions there.

additionally this would mean, that each message would have to contain all the possible parameters, unless we would write a bit more complicated serializer macro

we could signal which parameters are required by adding the non-default parameters to contructors

Reply:
I have removed the json header and direct usage of it from the pull request? I think now the issue should be resolved